ouroboros-consensus-protocol/src/ouroboros-consensus-protocol/Ouroboros/Consensus/Protocol/Ledger/HotKey.hs

```diff
@@ -157,66 +199,184 @@ sign = sign_
   -- | Max KES evolutions
   Word64 ->
   m (HotKey c m)
-mkHotKey initKey startPeriod@(Absolute.KESPeriod start) maxKESEvolutions = do
+mkHotKeyAtEvolution evolution ocert initKey startPeriod maxKESEvolutions =
+  mkHotKeyWith
+    (Just (ocert, initKey, evolution, startPeriod))
+    maxKESEvolutions
+    Nothing
+    (pure ())
+
+-- | Create a new 'HotKey' and initialize it to a poisoned state (containing no
+-- valid KES sign key).
+mkEmptyHotKey ::
+  forall m c.
+  (Crypto c, IOLike m) =>
+  -- | Max KES evolutions
+  Word64 ->
+  m () ->
+  m (HotKey c m)
+mkEmptyHotKey maxKESEvolutions =
+  mkDynamicHotKey maxKESEvolutions Nothing
+
+mkKESState ::
+  Word64 -> OCert.OCert c -> KES.SignKeyKES (KES c) -> Word -> Absolute.KESPeriod -> KESState c
+mkKESState maxKESEvolutions newOCert newKey evolution startPeriod@(Absolute.KESPeriod start) =
+  KESState
+    { kesStateInfo =
+        KESInfo
+          { kesStartPeriod = startPeriod
+          , kesEndPeriod = Absolute.KESPeriod (start + fromIntegral maxKESEvolutions)
+          , kesEvolution = evolution
+          }
+    , kesStateKey = KESKey newOCert newKey
+    }
+
+type KeyProducer c m =
+  -- | Callback that will be invoked when a new key has been received
+  (OCert.OCert c -> KES.SignKeyKES (KES c) -> Word -> Absolute.KESPeriod -> m ()) ->
+  -- | Callback that will be invoked when a key deletion has been received
+  m () ->
+  m ()
+
+-- | Create a new 'HotKey' that runs a key-producer action on a separate thread.
+-- The key producer action will receive a callback that can be used to pass
+-- keys into the HotKey; the HotKey will dynamically update its internal state
+-- to reflect new keys as they arrive.
+mkDynamicHotKey ::
+  forall m c.
+  (Crypto c, IOLike m) =>
+  -- | Max KES evolutions
+  Word64 ->
+  Maybe (KeyProducer c m) ->
+  m () ->
+  m (HotKey c m)
+mkDynamicHotKey = mkHotKeyWith Nothing
+
+-- | The most general function for creating a new 'HotKey', accepting an initial
+-- set of credentials, a key producer action, and a custom finalizer.
+mkHotKeyWith ::
+  forall m c.
+  (Crypto c, IOLike m) =>
+  Maybe (OCert.OCert c, KES.SignKeyKES (KES c), Word, Absolute.KESPeriod) ->
+  -- | Max KES evolutions
+  Word64 ->
+  Maybe (KeyProducer c m) ->
+  m () ->
+  m (HotKey c m)
+mkHotKeyWith initialStateMay maxKESEvolutions keyThreadMay finalizer = do
   varKESState <- newMVar initKESState
+
+  let set newOCert newKey evolution startPeriod =
+        modifyMVar_ varKESState $ \oldState -> do
+          _ <- poisonState oldState
+          return $ mkKESState maxKESEvolutions newOCert newKey evolution startPeriod
+      unset =
+        modifyMVar_ varKESState $ poisonState
+
+  forM_ initialStateMay $ \(newOCert, newKey, evolution, startPeriod) ->
+    set newOCert newKey evolution startPeriod
+
+  finalizer' <- case keyThreadMay of
+    Just keyThread -> do
+      keyThreadAsync <- async $ do
+        labelThisThread "HotKey receiver"
+        keyThread set unset
```

Comment:
This is convoluted, we spawn a new thread to which we give the `set` and `unset` functions, but I didn't find what that thread is supposed to do. Seems we always call this with `Nothing` as the `keyThreadMay`?